assets/src/components/v2/departures/section.ts

```diff
@@ -0,0 +1,121 @@
+import _ from "lodash";
+
+import NormalSection, { Layout } from "./normal_section";
+import NoticeSection from "./notice_section";
+
+export type Section =
+  | (NormalSection & { type: "normal_section" })
+  | (NoticeSection & { type: "notice_section" });
+
+/**
+ * Perform one iteration of "section trimming", i.e. the process of removing
+ * departures from sections while respecting their desired layouts until they
+ * all fit in a container without overflow.
+ *
+ * Operates immutably. If no sections could be trimmed, returns the same array.
+ */
+export const trimSections = (sections: Section[]): Section[] => {
+  const trimmed = _.cloneDeep(sections);
+
+  const sortedSizedSections = trimmed
+    .map((section) => ({ section, length: sectionLength(section) }))
+    .sort(({ length: a }, { length: b }) => b - a);
+
+  for (const key in trimStages) {
+    if (trimStages[key](sortedSizedSections)) return trimmed;
+  }
+
+  return sections;
+};
+
+// Determine how many "items" a section contains, for the purposes of deciding
+// which sections should be trimmed first.
+const sectionLength = (section: Section): number => {
+  if (section.type == "normal_section") {
+    return section.rows.reduce(
+      (sum, row) =>
+        sum +
+        (row.type == "departure_row" ? row.times_with_crowding.length : 1),
+      0,
+    );
+  } else {
+    return 0;
+  }
+};
+
+type SizedSection = {
+  length: number;
+  section: Section;
+};
+
+// Functions which destructively trim departures from sections, returning `true`
+// if any trimming occurred. Each iteration of section trimming tries each one
+// of these in sequence until one succeeds or they have all been tried. Assumes
+// the given sections are reverse-sorted by length (largest first).
+const trimStages: Record<string, (sections: SizedSection[]) => boolean> = {
+  // Trim all sections with more departures than their `max` until they have
+  // exactly as many as their `max` (if defined).
+  allToMax: (sections) => {
+    return sections.reduce((didTrim, { section, length }) => {
+      if (section.type == "normal_section" && section.layout.max) {
+        for (let l = length; l > section.layout.max; l--) {
+          if (trimSection(section)) {
+            // Trimmed this section; may be able to trim more, so keep going.
+            didTrim ||= true;
+          } else {
+            // This section can't be trimmed further, so return early.
+            return didTrim || false;
```

Comment:
**question:** is this the same as `return didTrim`?

Reply:
Hah, yep. I put the `true` and `false` here to correspond with the possible returns of `trimSection`, wanting to make sure it was OR'd with the accumulator in both cases, not thinking about the fact that `|| false` is a no-op.